modules/xfem/src/userobjects/MeshCut3DUserObject.C

```diff
@@ -0,0 +1,1106 @@
+
+  double plane_point[3] = {center(0), center(1), center(2)};
+  double planenormal[3] = {normal(0), normal(1), normal(2)};
+  double p_begin[3] = {p1(0), p1(1), p1(2)};
+  double p_end[3] = {p2(0), p2(1), p2(2)};
+
+  double cut_point[3] = {0.0, 0.0, 0.0};
+
+  if (Xfem::plane_normal_line_exp_int_3d(plane_point, planenormal, p_begin, p_end, cut_point) == 1)
+  {
+    Point p(cut_point[0], cut_point[1], cut_point[2]);
+    Real dotp = ((p - p1) * (p2 - p1)) / ((p2 - p1) * (p2 - p1));
+    if (isInsideCutPlane(vertices, p) && dotp > 1)
+    {
+      pint = p;
+      has_intersection = true;
+    }
+  }
+  return has_intersection;
+}
+
+bool
+MeshCut3DUserObject::isInsideEdge(const Point & p1, const Point & p2, const Point & p) const
+{
+  Real dotp1 = (p1 - p) * (p2 - p1);
+  Real dotp2 = (p2 - p) * (p2 - p1);
+  return (dotp1 * dotp2 <= 0.0);
+}
+
+Real
+MeshCut3DUserObject::getRelativePosition(const Point & p1, const Point & p2, const Point & p) const
+{
+  // get the relative position of p from p1
+  Real full_len = (p2 - p1).norm();
+  Real len_p1_p = (p - p1).norm();
+  return len_p1_p / full_len;
+}
+
+bool
+MeshCut3DUserObject::isInsideCutPlane(const std::vector<Point> & vertices, Point p) const
+{
+  unsigned int n_node = vertices.size();
+
+  Point normal;
+  Point center;
+
+  findElemNormal(vertices, normal, center);
+
+  bool inside = false;
+  unsigned int counter = 0;
+
+  for (unsigned int i = 0; i < n_node; ++i)
+  {
+    unsigned int iplus1 = (i < n_node - 1 ? i + 1 : 0);
+    Point middle2p = p - 0.5 * (vertices[i] + vertices[iplus1]);
+    const Point side_tang = vertices[iplus1] - vertices[i];
+    Point side_norm = side_tang.cross(normal);
+    Xfem::normalizePoint(middle2p);
+    Xfem::normalizePoint(side_norm);
+    if (middle2p * side_norm <= 0.0)
+      counter += 1;
+  }
+  if (counter == n_node)
+    inside = true;
+  return inside;
+}
+
+bool
+MeshCut3DUserObject::findElemNormal(const std::vector<Point> & vertices,
+                                    Point & normal,
+                                    Point & center) const
+{
+  unsigned int n_node = vertices.size();
+
+  center(0) = center(1) = center(2) = 0;
+  normal(0) = normal(1) = normal(2) = 0;
+
+  for (unsigned int i = 0; i < n_node; ++i)
+    center += vertices[i];
+  center /= n_node;
+
+  for (unsigned int i = 0; i < n_node; ++i)
+  {
+    unsigned int iplus1(i < n_node - 1 ? i + 1 : 0);
+    Point ray1 = vertices[i] - center;
+    Point ray2 = vertices[iplus1] - center;
+    normal += ray1.cross(ray2);
+  }
+  normal /= n_node;
+  Xfem::normalizePoint(normal);
+  return true; // to be done
+}
+
+void
+MeshCut3DUserObject::findBoundaryNodes()
+// this is a simple algorithm simply based on the added angle = 360 degrees
+// works fine for planar cutting surface
+// for curved cutting surface, need to re-work this subroutine to make it more general
+{
+  _boundary_map.clear();
+
+  unsigned int n_nodes = _cut_mesh->n_nodes();
+  Real angle[n_nodes];
```

Comment:
Avoid C-style arrays - `std::vector` is more appropriate here. Also you can assign zeros when you define it which means you don't need the explicit loop just below to initialize.